src/jot.gleam

```diff
@@ -99,71 +107,181 @@ fn drop_spaces(in: Chars) -> Chars {
   }
 }
 
+fn count_drop_spaces(in: Chars, count: Int) -> #(Chars, Int) {
+  case in {
+    [] -> #([], count)
+    [" ", ..rest] -> count_drop_spaces(rest, count + 1)
+    [c, ..rest] -> #([c, ..rest], count)
+  }
+}
+
 fn parse_document(
   in: Chars,
   refs: Refs,
   ast: List(Container),
   attrs: Dict(String, String),
 ) -> Document {
   let in = drop_lines(in)
-  let in = drop_spaces(in)
+  let #(in, spaces_count) = count_drop_spaces(in, 0)
+
+  let unknown_handler = fn(in, refs, ast, attrs) {
+    case in {
+      ["[", "^", ..in2] -> {
+        case parse_footnote_def(in2, "^") {
+          None -> {
+            let #(paragraph, in) = parse_paragraph(in, attrs)
+            parse_document(in, refs, [paragraph, ..ast], dict.new())
+          }
+          Some(#(id, footnote, in)) -> {
+            let refs =
+              Refs(..refs, footnotes: dict.insert(refs.footnotes, id, footnote))
+            parse_document(in, refs, ast, dict.new())
+          }
+        }
+      }
+      ["[", ..in2] -> {
+        case parse_ref_def(in2, "") {
+          None -> {
+            let #(paragraph, in) = parse_paragraph(in, attrs)
+            parse_document(in, refs, [paragraph, ..ast], dict.new())
+          }
+          Some(#(id, url, in)) -> {
+            let refs = Refs(..refs, urls: dict.insert(refs.urls, id, url))
+            parse_document(in, refs, ast, dict.new())
+          }
+        }
+      }
+      _ -> {
+        let #(paragraph, in) = parse_paragraph(in, attrs)
+        parse_document(in, refs, [paragraph, ..ast], dict.new())
+      }
+    }
+  }
+
+  parse_block_segment(
+    in,
+    refs,
+    ast,
+    attrs,
+    spaces_count,
+    parse_document,
+    fn(ast, refs, _) {
+      Document(
+        content: list.reverse(ast),
+        references: refs.urls,
+        footnotes: refs.footnotes,
+      )
+    },
+    unknown_handler,
+  )
+}
+
+fn parse_block(
+  in: Chars,
+  refs: Refs,
+  ast: List(Container),
+  attrs: Dict(String, String),
+  required_spaces: Int,
+) -> #(List(Container), Chars) {
+  let in = drop_lines(in)
+  let #(in, spaces_count) = count_drop_spaces(in, 0)
+
+  use <- bool.lazy_guard(spaces_count < required_spaces, fn() {
+    #(list.reverse(ast), in)
+  })
+  parse_block_indent_unchecked(
+    in,
+    refs,
+    ast,
+    attrs,
+    required_spaces,
+    spaces_count,
+  )
+}
+
+fn parse_block_indent_unchecked(
```

Comment:
What do these new functions do? With the "unknown handler" arguments. I'm not sure I understand the design yet. Some documentation would be useful, given they are more complex than before.

What does "unchecked" mean also?